[FLINK-33448][runtime] Introduce a new configuration item 'taskmanager.load-balance.mode'

RocMarshal opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Introduce a new configuration item 'taskmanager.load-balance.mode'   
The PR resulted from https://github.com/apache/flink/pull/23635#discussion_r1381212518


## Brief change log

Introduce a new configuration item 'taskmanager.load-balance.mode' and make it  compatible with "cluster.evenly-spread-out-slots"

The ticket is mainly to do three items:

- Introduce a new configuration item 'taskmanager.load-balance.mode'
- Make it  compatible with "cluster.evenly-spread-out-slots"
  - Only update `taskmanager.load-balance.mode` to Slots when the taskmanager.load-balance.mode isn't set and cluster.evenly-spread-out-slots is true.

- Mark "cluster.evenly-spread-out-slots" as Deprecated


## Verifying this change

This change is already covered by existing tests.

This change added tests and can be verified as follows:

- org.apache.flink.configuration.TaskManagerLoadBalanceModeTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

RocMarshal commented:
Hi, @1996fanrui Could you help take a look in your free time? Thank you~ :)

1996fanrui commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerLoadBalanceMode.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public static TaskManagerLoadBalanceMode loadFromConfiguration(
+            @Nonnull Configuration configuration) {
+        TaskManagerLoadBalanceMode taskManagerLoadBalanceMode =
+                configuration.get(TaskManagerOptions.TASK_MANAGER_LOAD_BALANCE_MODE);
+        boolean evenlySpreadOutSlots =
+                configuration.getBoolean(ClusterOptions.EVENLY_SPREAD_OUT_SLOTS_STRATEGY);
+
+        if (taskManagerLoadBalanceMode != TaskManagerLoadBalanceMode.NONE
+                && taskManagerLoadBalanceMode != null) {
+            return taskManagerLoadBalanceMode;
+        }
+        if (evenlySpreadOutSlots) {
+            return TaskManagerLoadBalanceMode.SLOTS;
+        }
```
> please note we only update `taskmanager.load-balance.mode` to Slots when the `taskmanager.load-balance.mode` isn't set and `cluster.evenly-spread-out-slots` is true.

As I metioned in FLINK-33348, the current code doesn't respect this rule.

- If user sets `taskmanager.load-balance.mode : None` and `cluster.evenly-spread-out-slots : true` manually, we should respect `taskmanager.load-balance.mode : None`.
- If user doesn't set `taskmanager.load-balance.mode : None` and set `cluster.evenly-spread-out-slots : true`, we should update `taskmanager.load-balance.mode` to `Slots`.

In biref, the default `None` isn't same with user set `taskmanager.load-balance.mode` to `None` manually.

How to check the None is default value or from user set? Please check `conf.getOptional()` instead of `conf.get()` , and please add a test for these 2 cases.

Please correct me if my understanding is wrong.

RocMarshal replied:
Sorry for not get correctly your viewpoint before. Thank you for your clarification

1996fanrui commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerLoadBalanceMode.java:
```diff
@@ -0,0 +1,54 @@
+
+/** Type of {@link TaskManagerOptions#TASK_MANAGER_LOAD_BALANCE_MODE}. */
+@Internal
+public enum TaskManagerLoadBalanceMode {
```
I don't think it should be a separate class, especially it shouldn't be marked as `@Internal` class, because users will use it during `conf.set`, such as : `conf.set(TASK_MANAGER_LOAD_BALANCE_MODE, TaskManagerLoadBalanceMode.SLOTS)`.

It's better to add the class as a internal class of `TaskManagerOptions`. Please check out `JobManagerOptions.SchedulerType`.

RocMarshal replied:
make sense~
updated.

1996fanrui commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerLoadBalanceMode.java:
```diff
@@ -0,0 +1,54 @@
+                configuration.getBoolean(ClusterOptions.EVENLY_SPREAD_OUT_SLOTS_STRATEGY);
+
+        if (taskManagerLoadBalanceMode != TaskManagerLoadBalanceMode.NONE
+                && taskManagerLoadBalanceMode != null) {
```
You wrote a couple of `taskManagerLoadBalanceMode != null` or `taskManagerLoadBalanceMode == null`.

IIUC, it never be null due to `TaskManagerOptions.TASK_MANAGER_LOAD_BALANCE_MODE` has default value.

Please clean up your code and remove uncessary code, thanks~

RocMarshal replied:
Thanks for the reminding~ 
I checked the value-passing logic. It's precise on your comments.

1996fanrui commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -708,6 +708,23 @@ public class TaskManagerOptions {
+                                            "Mode for the load-balance allocation strategy across all available %s. "
+                                                    + "The %s mode tries to spread out the slots evenly across all available %s. "
+                                                    + "The %s mode is the default mode without any specified strategy. ",
+                                            code("TaskExecutors"),
```
Would you mind updating it from `TaskExecutors` to `TaskManagers`?

TaskExecutor is a class. From user side, we always call it `TaskManagers`.

1996fanrui commented on flink-core/src/test/java/org/apache/flink/configuration/TaskManagerLoadBalanceModeTest.java:
```diff
@@ -0,0 +1,62 @@
+        assertThat(TaskManagerLoadBalanceMode.loadFromConfiguration(conf))
+                .isEqualTo(TaskManagerLoadBalanceMode.SLOTS);
+
+        // Check for 'taskmanager.load-balance.mode: SLOTS' and 'cluster.evenly-spread-out-slots:
+        // true'
+        conf.setString(EVENLY_SPREAD_OUT_SLOTS_STRATEGY.key(), "true");
+        assertThat(TaskManagerLoadBalanceMode.loadFromConfiguration(conf))
+                .isEqualTo(TaskManagerLoadBalanceMode.SLOTS);
```
It's better to split these tests to separate code block. The conf is used and changed by all cases, it's not clear.

For example, you want to test `taskmanager.load-balance.mode: SLOTS and cluster.evenly-spread-out-slots: false` in the third case, you must check what the `cluster.evenly-spread-out-slots` is in the second case. It's not clear and easy to bug.


```
// code block1
{
  Configuration conf = new Configuration();
}


// code block2
{
  Configuration conf = new Configuration();
}
```

RocMarshal replied:
Sounds nice to me.
updated~

1996fanrui reviewed (changes requested):
Thanks @RocMarshal for your quick contribution! Great work!

I left a series of comments, please take a look in your free time, thanks~

1996fanrui commented on flink-core/src/test/java/org/apache/flink/configuration/TaskManagerLoadBalanceModeTest.java:
```diff
@@ -0,0 +1,62 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for {@link TaskManagerLoadBalanceMode}. */
+@ExtendWith(TestLoggerExtension.class)
```
```suggestion
```

`org.apache.flink.util.TestLoggerExtension` has been loaded fort the whole `flink-core` module, so the class level doesn't need it.

RocMarshal commented on flink-core/src/main/java/org/apache/flink/configuration/ClusterOptions.java:
```diff
@@ -84,17 +88,47 @@ public class ClusterOptions {
+                                                            .key()),
+                                            code("NONE"),
+                                            code(EVENLY_SPREAD_OUT_SLOTS_STRATEGY_KEY),
+                                            code("false"))
```
updated~

RocMarshal commented:
Hi, @1996fanrui Thanks a lot for the comments. I addressed the comments based on your suggestion.
Would you @zhuzhurk @KarmaGYZ @1996fanrui please mind helping take a look on it  if you have free time? 
Thank you!

1996fanrui commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -708,6 +711,49 @@ public class TaskManagerOptions {
+
+            return evenlySpreadOutSlots
+                    ? TaskManagerLoadBalanceMode.SLOTS
+                    : TaskManagerLoadBalanceMode.NONE;
```
```suggestion
                    : TaskManagerOptions.TASK_MANAGER_LOAD_BALANCE_MODE.defaultValue();
```

1996fanrui commented on flink-core/src/test/java/org/apache/flink/configuration/TaskManagerLoadBalanceModeTest.java:
```diff
@@ -0,0 +1,76 @@
+        // 'cluster.evenly-spread-out-slots: false'
+        Configuration conf1 = new Configuration();
+        assertThat(TaskManagerLoadBalanceMode.loadFromConfiguration(conf1))
+                .isEqualTo(TaskManagerLoadBalanceMode.NONE);
```
```suggestion
                .isEqualTo(TASK_MANAGER_LOAD_BALANCE_MODE.defaultValue());
```

1996fanrui commented on flink-core/src/test/java/org/apache/flink/configuration/TaskManagerLoadBalanceModeTest.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Test
+    void testReadTaskManagerLoadBalanceMode() {
+        // Check for non-set 'taskmanager.load-balance.mode: NONE' and
```
```suggestion
        // Check for non-set 'taskmanager.load-balance.mode' and
```

1996fanrui commented on flink-core/src/test/java/org/apache/flink/configuration/TaskManagerLoadBalanceModeTest.java:
```diff
@@ -0,0 +1,76 @@
+        assertThat(TaskManagerLoadBalanceMode.loadFromConfiguration(conf1))
+                .isEqualTo(TaskManagerLoadBalanceMode.NONE);
+
+        // Check for non-set 'taskmanager.load-balance.mode: NONE' and
```
```suggestion
        // Check for non-set 'taskmanager.load-balance.mode' and
```

1996fanrui reviewed (commented):
Hi @RocMarshal , thanks for your quick update!

Overall the latest code looks good to me and I left some small comments.

KarmaGYZ commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -708,6 +711,49 @@ public class TaskManagerOptions {
                             "Time we wait for the timers in milliseconds to finish all pending timer threads"
                                     + " when the stream task is cancelled.");
 
+    @Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER)
```
Should it belong to `EXPERT_SCHEDULING`?

RocMarshal replied:
big +1~

KarmaGYZ commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -708,6 +711,49 @@ public class TaskManagerOptions {
+                            Description.builder()
+                                    .text(
+                                            "Mode for the load-balance allocation strategy across all available %s. "
+                                                    + "The %s mode tries to spread out the slots evenly across all available %s. "
+                                                    + "The %s mode is the default mode without any specified strategy. ",
```
You can use the `list()` to list all the valid options.

KarmaGYZ commented on flink-core/src/test/java/org/apache/flink/configuration/TaskManagerLoadBalanceModeTest.java:
```diff
@@ -0,0 +1,76 @@
+
+        // Check for setting manually 'taskmanager.load-balance.mode: NONE' and
+        // 'cluster.evenly-spread-out-slots: true'
+        Configuration conf4 = new Configuration();
+        conf4.set(TASK_MANAGER_LOAD_BALANCE_MODE, TaskManagerLoadBalanceMode.NONE);
+        conf4.set(EVENLY_SPREAD_OUT_SLOTS_STRATEGY, true);
+        assertThat(TaskManagerLoadBalanceMode.loadFromConfiguration(conf4))
+                .isEqualTo(TaskManagerLoadBalanceMode.NONE);
```
Following the FLIP-321, I think we need to guard the backward compatibility here. If a user configures the `EVENLY_SPREAD_OUT_SLOTS_STRATEGY`, the behavior should be equal to configuring `TaskManagerLoadBalanceMode.SLOTS`.

RocMarshal replied:
Hi, @KarmaGYZ Thanks for the comment.
I guess your concern should be the backward compatibility handling of cases like Case1 and Case2 between `line-33-line44`?   Because in the current case, if the user proactively sets this value, they should understand the meaning of this option. In other words, in this case, `None` is the final value expected by the user. 
Please let me know what's your opinion~ :)

KarmaGYZ replied:
sounds good to me.

KarmaGYZ reviewed (approved):
+1 for merging.

1996fanrui reviewed (approved):
Thanks @RocMarshal  for the contribution, and @KarmaGYZ for the review.

Hi @zhuzhurk , I will merge this PR first to move forward due to this PR is simple enough and the subsequent PR is waiting for this PR. Welcome your feedback if you have any suggestion later, @RocMarshal will update them in the subsequent PR, thanks~

zhuzhurk commented:
Sorry that I was on a trip last week and did not have a chance to take a look. The change looks good to me. Just one minor comment that it's better to annotate `TaskManagerLoadBalanceMode#loadFromConfiguration()` with `@Internal` to avoid indicating it as a public interface to users.

1996fanrui commented:
> Sorry that I was on a trip last week and did not have a chance to take a look. The change looks good to me. Just one minor comment that it's better to annotate `TaskManagerLoadBalanceMode#loadFromConfiguration()` with `@Internal` to avoid indicating it as a public interface to users.

Thanks @zhuzhurk for the review and good suggestion!

cc @RocMarshal

RocMarshal commented:
Thanks  @zhuzhurk for the comment & @1996fanrui for the ping.
I'll drive a simple jira & pr to advance it.

1996fanrui commented:
> Thanks @zhuzhurk for the comment & @1996fanrui for the ping. I'll drive a simple jira & pr to advance it.

@RocMarshal No, don't need extra PR and JIRA. As @KarmaGYZ 's suggestion, it's better to finish all commits related to DefaultSchduler of in #23635 . And merge them after all commits are perfect. Otherwise, it will introduce a lot of minor fix commits, such as this suggestion[1].

[1] https://github.com/apache/flink/pull/23658#issuecomment-1809516092
